src/EntityComponentManager.cc

```diff
@@ -795,27 +802,32 @@ const EntityGraph &EntityComponentManager::Entities() const
 }
 
 //////////////////////////////////////////////////
-detail::BaseView *EntityComponentManager::FindView(
+std::pair<detail::BaseView *, std::mutex *> EntityComponentManager::FindView(
     const std::vector<ComponentTypeId> &_types) const
 {
   std::lock_guard<std::mutex> lockViews(this->dataPtr->viewsMutex);
+  std::pair<detail::BaseView *, std::mutex *> viewMutexPair(nullptr, nullptr);
   auto iter = this->dataPtr->views.find(_types);
   if (iter != this->dataPtr->views.end())
-    return iter->second.get();
-  return nullptr;
+  {
+    viewMutexPair.first = iter->second.first.get();
+    viewMutexPair.second = iter->second.second.get();
+  }
+  return viewMutexPair;
 }
 
 //////////////////////////////////////////////////
 detail::BaseView *EntityComponentManager::AddView(
     const detail::ComponentTypeKey &_types,
     std::unique_ptr<detail::BaseView> _view) const
 {
-  // If the view already exists, then the map will return the iterator to
-  // the location that prevented the insertion.
   std::lock_guard<std::mutex> lockViews(this->dataPtr->viewsMutex);
-  auto iter = this->dataPtr->views.insert(
-      std::make_pair(_types, std::move(_view))).first;
-  return iter->second.get();
+  auto iter = this->dataPtr->views.find(_types);
+  if (iter != this->dataPtr->views.end())
+    return iter->second.first.get();
+  this->dataPtr->views[_types] =
+    std::make_pair(std::move(_view), std::make_unique<std::mutex>());
+  return this->dataPtr->views[_types].first.get();
 }
```

Comment:
I tried to modify the `this->dataPtr->views.insert` call to pass in a `std::pair<detail::ComponentTypeKey, std::pair<std::unique_ptr<detail::BaseView>, std::unique_ptr<std::mutex>>`, but was having trouble getting it to compile. So, I resorted to the changes made here (searching and then manually adding if the search didn't find anything). I don't think this is a bad change though since the `AddView` method is only called [here](https://github.com/ignitionrobotics/ign-gazebo/blob/166c06580036943c62d49422c3ef235b4eb7ce5d/include/ignition/gazebo/detail/EntityComponentManager.hh#L549), and the changes made here introduce only an extra lookup or two in a `std::unordered_map`, which already has `O(1)` lookup time. Since the `AddView` method is only called when a requested view doesn't already exist, this means that this code should rarely be called anyways, so I don't think the extra lookups are a problem.

Reply:
is this something that you're thinking of doing?

```
  auto iter = this->dataPtr->views.insert(
    std::make_pair(_types,
    std::make_pair(std::move(_view), std::make_unique<std::mutex>()))).first;
  return iter->second.first.get();
```

Reply:
Yeah, that's it! I tried what you suggested above, and it works. I think the issue I was having before was that I made the view/mutex pair outside of the `insert` call, and then the compiler was complaining to me about trying to copy a `unique_ptr`. I went ahead and made the update in https://github.com/ignitionrobotics/ign-gazebo/pull/1001/commits/395495816478b6aa102dc807b6e8b2d09f291ea3.